Flat Cosine Annealing

muellerzr opened the pull request:
This PR has the Flat Cosine Annealing Scheduler and fit function that should be used for the new optimizers such as RAdam and Ranger. PR's from those should be soon. See discussion here for relevant information about the "competition" held: [url](https://forums.fast.ai/t/how-we-beat-the-5-epoch-imagewoof-leaderboard-score-some-new-techniques-to-consider/53453)

Documentation to come soonish. If it is easier to keep it as one PR then it can wait

sgugger commented on fastai/callbacks/flat_cos_anneal.py:
```diff
@@ -0,0 +1,64 @@
+            curve_type=annealing_exp
+        else:
+            raiseValueError(f"annealing type not supported {curve}")
+        phase0 = TrainingPhase(self.anneal_start).schedule_hp('lr', lr).schedule_hp('mom', moms[0])
```
Just wondering, since you're using the phases API, can't we just create a function that returns a `GeneralScheduler`?

muellerzr replied:
Yes, I will look into it

muellerzr commented:
I believe I simplified it per your request. The callback should just have an __init__ to set up the scheduler correct?

muellerzr commented:
Fixes are all done, let me know if that's close to what you are talking about via the API

sgugger commented on fastai/callbacks/flat_cos_anneal.py:
```diff
@@ -0,0 +1,37 @@
+# The brain child of Mikhail Grankin aimed for use of the new optimizers
+# For more information: https://forums.fast.ai/t/how-we-beat-the-5-epoch-imagewoof-leaderboard-score-some-new-techniques-to-consider/53453
+
+class FlatCosAnnealScheduler(LearnerCallback):
```
I think you should make this a regular function that returns `GeneralScheduler(learn, phases)`. Keep the name capitalized for consistency with the other callbacks.

Second comment: please keep the docstring to one line and add any other info in the doc notebooks when you add the documentation. The link and citation to you are already in the source code in the commented lines 9 and 10, so we don't need it here. You should absolutely put them again in the documentation of this when you make that second PR.

muellerzr replied:
Adjusted it so.

sgugger commented:
We are on the right track. I think we can simplify it a little bit more.

Also, I have no idea why that is, but the diff on the train.py file shows the entire file as changed, not just the things you added. Do you have any idea why that is? It makes reviewing the PR a bit hard for me.

muellerzr commented:
I looked into it and tried a few things but I have absolutely no idea why. At first I thought it was because I had moved `fit_fc` to the start, but that did not do anything when I had moved it to the very end. I simplified it a bit more for you, and also got rid of the scheduler function as I don't believe that would've been beneficial here since it's just a function now. But let me know if we should have kept it.

lgrz commented:
> We are on the right track. I think we can simplify it a little bit more.
> 
> Also, I have no idea why that is, but the diff on the train.py file shows the entire file as changed, not just the things you added. Do you have any idea why that is? It makes reviewing the PR a bit hard for me.

FYI, it was because <LF> was replaced with <CRLF> (windows line endings)
